db/keyval/filedb/client.go

```diff
@@ -0,0 +1,422 @@
+	for _, filePath := range filePaths {
+		if dc := c.getFileDecoder(filePath); dc != nil {
+			byteFile, err := c.fsh.ReadFile(filePath)
+			if err != nil {
+				return nil, errors.Errorf("failed to read file %s content: %v", filePath, err)
+			}
+			fileEntries, err := dc.Decode(byteFile)
+			file := &decoder.File{Path: filePath, Data: fileEntries}
+			if err != nil {
+				return nil, errors.Errorf("failed to decode file %s: %v", filePath, err)
+			}
+			files = append(files, file)
+		}
+	}
+	// Put all the configuration for given agent to the database
+	for _, file := range files {
+		for _, data := range file.Data {
+			if ok := c.checkAgentPrefix(data.Key); ok {
+				c.db.Add(file.Path, &decoder.FileDataEntry{Key: data.Key, Value: data.Value})
+			}
+		}
+	}
+	// Validate and prepare the status file and decoder
+	if c.stPath != "" {
+		c.stDc = c.getFileDecoder(c.stPath)
+		if c.stDc == nil {
+			return nil, errors.Errorf("failed to get decoder for status file (unknown extension) %s: %v", c.stPath, err)
+		}
+		filePath, err := c.fsh.GetFileNames([]string{c.stPath})
+		if err != nil {
+			return nil, errors.Errorf("failed to read status file: %v", err)
+		}
+		// Expected is at most single entry
+		if len(filePath) == 0 {
+			if err := c.fsh.CreateFile(c.stPath); err != nil {
+				return nil, errors.Errorf("failed to create status file: %v", err)
+			}
+		} else if len(filePath) > 1 {
+			return nil, errors.Errorf("failed to process status file, unexpected processing output: %v", err)
+		}
+	}
+
+	return c, nil
+}
+
+// GetPaths returns client file paths
+func (c *Client) GetPaths() []string {
+	return c.cfgPaths
+}
+
+// GetPrefix returns current agent prefix
+func (c *Client) GetPrefix() string {
+	return c.agentPrefix
+}
+
+// GetDataForFile returns data gor given file
+func (c *Client) GetDataForFile(path string) []*decoder.FileDataEntry {
+	return c.db.GetDataForFile(path)
+}
+
+// GetDataForKey returns data gor given file
+func (c *Client) GetDataForKey(key string) (*decoder.FileDataEntry, bool) {
+	return c.db.GetDataForKey(key)
+}
+
+// Represents data with both identifiers, file path and key.
+type keyedData struct {
+	path string
+	watchResp
+}
+
+// BrokerWatcher implements CoreBrokerWatcher and provides broker/watcher constructors with client
+type BrokerWatcher struct {
+	*Client
+	prefix string
+}
+
+// NewBroker provides BytesBroker object with client and given prefix
+func (c *Client) NewBroker(prefix string) keyval.BytesBroker {
+	return &BrokerWatcher{
+		Client: c,
+		prefix: prefix,
+	}
+}
+
+// NewWatcher provides BytesWatcher object with client and given prefix
+func (c *Client) NewWatcher(prefix string) keyval.BytesWatcher {
+	return &BrokerWatcher{
+		Client: c,
+		prefix: prefix,
+	}
+}
+
+// Put reads status file, add data to it and performs write
+func (c *Client) Put(key string, data []byte, opts ...datasync.PutOption) error {
+	// Read and decode status file
+	stFile, err := c.fsh.ReadFile(c.stPath)
```

Comment:
This will need optimization, imagine 100x Put.. every time file needs to be read from disk, decoded, looped through, encoded, written to disk. That is too much overhead. Some caching should be introduced at the filesystem handler to prevent unnecessary re-reading.

If not implemented in this PR, at least some TODO should be mentioned here.

Reply:
Since the database is in sync with the filesystem, data do not need to be actually read from disk.